assembly/src/parser/mod.rs

```diff
@@ -0,0 +1,159 @@
+/// Simple macro used in the grammar definition for constructing spans
+macro_rules! span {
+    ($l:expr, $r:expr) => {
+        crate::SourceSpan::new($l..$r)
+    };
+    ($i:expr) => {
+        crate::SourceSpan::new($i..$i)
+    };
+}
+
+lalrpop_util::lalrpop_mod!(
+    #[allow(clippy::all)]
+    grammar,
+    "/parser/grammar.rs"
+);
+
+mod error;
+mod lexer;
+mod location;
+mod scanner;
+mod span;
+mod token;
+
+pub use self::error::{HexErrorKind, LiteralErrorKind, ParsingError};
+pub use self::lexer::Lexer;
+pub use self::location::SourceLocation;
+pub use self::scanner::Scanner;
+pub use self::span::{SourceSpan, Span, Spanned};
+pub use self::token::{DocumentationType, HexEncodedValue, Token};
+
+use crate::{
+    ast,
+    diagnostics::{Report, SourceFile},
+    sema, LibraryPath,
+};
+use alloc::{boxed::Box, collections::BTreeSet, sync::Arc, vec::Vec};
+
+type ParseError<'a> = lalrpop_util::ParseError<u32, Token<'a>, ParsingError>;
+
+// MODULE PARSER
+// ================================================================================================
+
+#[derive(Default)]
+pub struct ModuleParser {
+    kind: ast::ModuleKind,
+    interned: BTreeSet<Arc<str>>,
+}
```

Comment:
Could we add some doc comments to this struct. For example, it wasn't clear to me what exactly is `interned` is for.